Support some presentation remotes

moreau-nicolas opened the pull request:
Some presentation remotes use Left/Right to navigate the slides.
By default Reveal uses those for column/chapter navigation.
This is far less useful than slide navigation.

This patch lets the presenter use Left/Right to navigate slides, but
still Does The Right Thing in overview mode.

We also map the Up/Down arrows for consistency.

moreau-nicolas committed:
Support some presentation remotes

Some presentation remotes use Left/Right to navigate the slides.
By default Reveal uses those for column/chapter navigation.
This is far less useful than slide navigation.

This patch lets the presenter use Left/Right to navigate slides, but
still Does The Right Thing in overview mode.

We also map the Up/Down arrows for consistency.

hgwood commented:
Nice idea. However it means navigation between chapters with left/right will not be possible anymore, and that might disturb quite a few presenters. How about enabling the bindings you're proposing conditionally with a query parameter? Maybe `?remote`? What do you think?

moreau-nicolas commented:
It is in fact still possible to skip chapters, by using Reveal's "overview mode".

I'm not sure how I feel about using a query parameter. I'll think it over. :)

moreau-nicolas committed:
Add a query parameter

The `?remote` query parameter now enables the "remote mode" whic remaps
some key bindings.

moreau-nicolas committed:
Add key binding for "remote mode"

The [R] key now toggles the "remote mode".
It does this by updating the `window.location.search` attribute.

moreau-nicolas commented:
@hgwood I added a query parameter for feature toggling (as you suggested).

I also added a key binding in order to have more a user-friendly way to enable the feature (yes, I'm that lazy :D).

I'm unsure as where to document this new feature. The Wiki ? The README ? Somewhere else ?

hgwood commented on reveal/run.js:
```diff
@@ -73,6 +74,37 @@
+	.replace(/[?]remote[&]/i, '?')
+        .replace(/[?]remote/i, '')
+	.replace(/[&]remote[&]/i, '&')
+        .replace(/[&]remote$/i, '');
```
Looks like some lines are missing the indentation.

hgwood commented on reveal/run.js:
```diff
@@ -73,6 +74,37 @@
+    if (window.location.search.match(/[?]/))
+      window.location.search = window.location.search + '&remote';
+    else
+      window.location.search = '?remote';
```
Could you add some braces there and lines 103-106? It's fine to omit braces when condition and consequence are on the same line but here it might trip the next contributor up.

hgwood replied:
Actually I think a ternary would be even better here.

hgwood commented on reveal/run.js:
```diff
@@ -35,6 +35,13 @@
+        39: function rightArrow() { if (isRemoteMode()) Reveal.next(); else Reveal.right(); },
+        40: function downArrow() { if (isRemoteMode()) Reveal.next(); else Reveal.down(); },
+        82: function rKey() { toggleRemoteMode(); },
+      },
```
Maybe this could be a little less dense? What about this:

```js
37: function leftArrow() {
  if (isRemoteMode()) Reveal.prev();
  else Reveal.left();
},
```

hgwood commented on reveal/run.js:
```diff
@@ -67,6 +74,37 @@
+    if (hasRemoteQueryParameter())
+      disableRemoteMode();
+    else
+      enableRemoteMode();
```
Maybe this would a bit less dangerous for the next contributor:

```js
if (...) ...;
else ...;
```

hgwood reviewed (changes requested):
Thanks! Only stylisting issues. It's probably time to add eslint to this project. :)

moreau-nicolas committed:
Stylistic changes

moreau-nicolas commented:
@hgwood please review :)

hgwood commented on reveal/run.js:
```diff
@@ -36,11 +36,25 @@
+            },
+        82: function rKey() {
+              toggleRemoteMode();
+            },
```
I'm nitpicking, but this does not look like conventional JS indentation. It should look like this:

```js
82: function rKey() {
  toggleRemoteMode();
},
```

hgwood commented on reveal/run.js:
```diff
@@ -85,25 +99,22 @@
-      window.location.search = '?remote';
+    window.location.search += window.location.search.match(/[?]/)
+                           ? '&remote'
+                           : "?remote";
```
- Such alignement are difficult to maintain (eg if the variable being assigned changes). Lets leave conventional indentation (ie the last two lines should be indented one more step from the first one).
- The last line should use single quotes, like the rest of the code.

hgwood reviewed (changes requested):
Only styling issues.

hgwood commented:
@moreau-nicolas I was on leave this week :)

moreau-nicolas committed:
Implement requested changes

moreau-nicolas commented:
@hgwood hope you had a great time :)

hgwood commented:
Thanks for your contribution and patience. :)

zigarn commented:
@moreau-nicolas : Could you check if this need to be ported on sensei or it's not necessary with RevealJS 4?

moreau-nicolas commented:
Just tried the sample on revealjs.com (using 4.1.0) and it does behave the same as older RevealJS releases.
This definitely should be ported.
I'll do it.
